### PR TITLE
Added Copy Permalink Functionality to Comment Dropdown

### PR DESCRIPTION
**Pull request by nizarmah**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR adds a *Copy Permalink* button to the dropdown list.

Copy functionality is similar to the way 'copy article link' is implemented. It checks for the android bridge first, then the navigator, then fallsback to the default document copy command.

Since the codes were really similar, I moved the copy functionality to a function of its own with callbacks and fallbacks as lambdas, and triggered it with the copyArticleLink and the copyPermalink.

This PR is needed for the feature request on the [DEV-Android](https://github.com/thepracticaldev/DEV-Android/issues/66). Two suggestions were proposed, with the help of @fdoxyz:
- Copying the Permalink instead of redirecting when pressing the *Permalink* anchor tag.
- Creating a *Copy Permalink*.

And here we are :smile: 

## Related Tickets & Documents
thepracticaldev/DEV-Android#66

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
image
image

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## What gif best describes this PR or how it makes you feel?

hell yeaaaaaah

### CODE REVIEW

**Comment by nizarmah** [comment]
The PR is currently `[WIP]` because I'm checking if tests are working or not. Once I do that, I'll check if there is any need to add new tests or not.

**Comment by fdoxyz** [comment]
Thank you again for the PR @nizarmah 🙌🏼 I'll take a look at this today

**Comment by nizarmah** [comment]
Always my pleasure @fdoxyz ( :

Whenever it is possible, let me know if the PR needs any Test cases. If it doesn't, so I change it from draft to ready.

**Comment by fdoxyz** [comment]
Also about the question on testing: I think it may be valuable to test the `copyText` function since we are now reusing it.

We can have a request (integration) test for this. In case you're not comfortable with these (they are Ruby specs) I can help push a commit with it. However, there may be a better way to do this without using request specs. 

I'd say for now once the small issues are fixed we can open up this PR for review and get feedback from others. In the mean time I'll ask for help to other core contributors about the best strategy to test this.

**Comment by nizarmah** [comment]
@fdoxyz replaced the callbacks with the promises. **I'm really sorry that the replies are taking so long on this pull request.**

Anyway, I don't know any Ruby, and I don't have any idea on how to organize the writing for the javascript test, incase it ended up being a unit test and not an integration one. Maybe then we can test the Android Bridge and the navigator separately too. But I'd be more than happy to learn and apply it :smile: 

Also, can you please clear the cache for the dropdown? Personally for me, after increasing the width to `138px` it didn't overflow outside the dropdown. I had the same issue when my cache wasn't cleared though. Please let me know :slightly_smiling_face:

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: I'm glad :slightly_smiling_face: Yeah, regarding the Promises, I had no idea about them. I kind of stopped JavaScript before I got introduced to them, but thanks a lot for introducing me to them :+1: They'll for sure come in handy later on.

**Comment by fdoxyz** [comment]
No worries @nizarmah! We've come a long way to get this feature in for mobile and your help is greatly appreciated. In this repo we'll need approval from other contributors before we can merge but I'm hoping we're not far from that.

I cleared my cache and I still had the same problem, which is a bit weird, not sure if it's the OS/browser. I tried changing the `button` tags back to `a` tags and that worked for me (no more text wrapping).

I just shared these small changes with a PR to your branch let me know what you think there. Can you also please sync with master so it's closer to where we are currently. That way we can make sure there's no issues with the changes introduced this week :)

**Comment by nizarmah** [comment]
Now the _Copy Permalink_ anchor tag looks as follows (due to the merge with master)

image

**Comment by nizarmah** [comment]
@fdoxyz should I consider this PR ready for review now? Or shall I try to write some tests first?

**Comment by fdoxyz** [comment]
I'm still on the fence about using integration tests because it may be tricky/overkill, unsure in the end what would be the best way to do it.

That's why I think it's probably better to get feedback on it first and see what others may think is the best way to test this (or if even necessary in the end). I'm opening up for reviews now 😃

**Comment by nizarmah** [comment]
@fdoxyz any "internal" updates on this?

**Comment by fdoxyz** [comment]
Sorry for the delay in getting back around to this and thanks for the reminder @nizarmah !

I'm going to bump this again so we can hopefully hear feedback from others on this 👍

**Comment by joshpuetz** [comment]
This seems like a good workaround short of actually adding a system level share button in the mobile app's frame/chrome.

Question: does explicitly support for iOS need to be added to this as well?

**Comment by nizarmah** [comment]
@joshpuetz In case this question is directed towards me and not @fdoxyz, I don't think any iOS support will be needed. 

The `AndroidBridge` support is an exception to an issue caused by Android `WebView`. 

Meanwhile, I checked [DEV-ios](https://github.com/thepracticaldev/DEV-ios/) for any previous clipboard issues and I found none. Since I didn't change any functionality in `copyText` but parameterized (*can I even say that?*) the copied text, I don't think any extra iOS support is necessary. 

Also, I believe @fdoxyz tested this on iOS and everything should be working fine :+1:

If the question isn't directed to me, then please ignore this :slightly_frowning_face: 
Otherwise, I hope it answers your question :slightly_smiling_face:

**Comment by fdoxyz** [comment]
That's correct as @nizarmah mentioned: The "Copy to clipboard" functionality is supported in iOS from within the WebViews. The native bridge functionality is in place to workaround Android's WebView lack of support for this (using the native API for "copy to clipboard").

> This seems like a good workaround short of actually adding a system level share button in the mobile app's frame/chrome.

That is very true, we do have a way to do a "native share" functionality on Android ([thanks to a recently merged contribution](https://github.com/thepracticaldev/DEV-Android/pull/61)) and haven't developed this for iOS yet. This is separate/different than just "copy to clipboard" because you can open the text with any other app (Facebook, Twitter, etc).

Reworking and taking better advantage of "share" capabilities on both platforms is an idea I pitched a while ago but just hasn't been a top priority yet. I thought it would be best to have some thought put into this as this could benefit from the product/design team feedback as well.

Anyways, I see two options to go forward since that we're considering this:
1. Go ahead with this PR as "Copy to clipboard" and later on add a "Share" feature using the native capabilites on both platforms (this works for both Desktop browser + mobile)
1. Rename "Copy to clipboard" to "Share" and have it call `AndroidBridge.shareText` instead as Android's share feature allows for copy to clipboard (this is not released in production yet so it may have to wait a bit more)
1. Other ideas?

__The caveat here__ is that we would want to have "Share" not visible in the Desktop browser (only mobile), because this is a permalink and it would mean we have to refactor to include all the Desktop browser share functionalities (see images below that show all desktop sharing capabilities). Also, the second option feels a bit more like a patch instead of a properly shaped "share support" for all platforms, __which is why I'm personally leaning more towards going for the first option__ (unless other ideas are suggested)

----------

Screen Shot 2020-08-13 at 13 09 21

### Article share capabilities on Desktop

----------

Screen Shot 2020-08-13 at 13 10 39

### Comment share capabilities on Desktop

**Comment by joshpuetz** [comment]
Thanks for that explanation @nizarmah!

I agree with @fdoxyz that option 1 (going ahead with the PR, looping back to add proper native mobile OS 'share' support later) is the best choice to get this lack of functionality on Android resolved as soon as possible!

**Comment by nizarmah** [comment]
@fdoxyz any updates on when this PR would be reviewed? I've been keeping an eye on the merge conflicts and resolving them in the meantime :slightly_smiling_face:

**Comment by fdoxyz** [comment]
@nizarmah thank you for keeping the PR up to date. I'm really looking forward to see your contribution merged, I think you've done an all-around great job here 👏🏼

Back when we started talking about this feature (a while ago) we had "copy to clipboard" code spread throughout the codebase. In the mean time on other (unrelated) PRs we've slowly shaped an idea of `Runtime` to help abstract the support for Android and other mobile platforms too.

Other contributors are also working on "copy to clipboard" features in different areas of the codebase so I just sent in a PR to help unblock you both, in the cleanest way possible. Once #10323 is merged I'd ask you to refactor a bit one last time so the code can just execute `Runtime.copyToClipboard(____)`.

With that in place I'd surely give this an approval and hope to get a second approval from another team member (as that's the policy in place for merging). Thank you again for the help and patience!

**Comment by nizarmah** [comment]
@fdoxyz hello :D I noticed the PR was updated so I fixed the merge conflicts. 

It seems like the Jenkins Test failed at something I didn't touch/change. How would I be able to re-run the tests for Jenkins?

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializeCommentDropdown.js:35]: We can take advantage of the Snackbar to give users visual feedback here (catch to avoid unexpected problem if the Snackbar happens to not be available).

```javascript
Runtime.copyToClipboard(permalink).then(() => {
   addSnackbarItem({ message: 'Copied to clipbaord' });
}).catch((e) => { console.log(e); });
```

**Review by fdoxyz — CHANGES_REQUESTED** [review]
Awesome thank you @nizarmah, the changes are looking very good! Yeah CI sometimes runs in some hiccups I can restart the build for you.

However, while testing this I found out that my PR introduced a bug in the comment permalink page. Would you mind including this change in your PR? The fix consists of making the `shouldCloseDropdown` funtion from [initializeCommentDropdown.js](app/assets/javascripts/initializers/initializeCommentDropdown.js) look like this (currently when `copyIcon` isn't available it fails):

```js
function shouldCloseDropdown(event) {
    var copyIcon = document.getElementById('article-copy-icon');
    var isCopyIconChild = copyIcon && copyIcon.contains(event.target);
    return !(
      event.target.matches('.dropdown-icon') ||
      event.target.matches('.dropbtn') ||
      event.target.matches('clipboard-copy') ||
      isCopyIconChild ||
      event.target.parentElement.classList.contains('dropdown-link-row')
    );
  }
```

Also leaving another inline comment that would make this look just a little better (some visual feedback). As soon as you push another commit Travis should re-run the tests, if the problem is still there I'll re-run manually.

So close! And thanks again for sticking through with this 🙌🏼

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [inline app/assets/javascripts/initializers/initializeCommentDropdown.js:35]:
```suggestion
Runtime.copyToClipboard(permalink).then(() => {
    addSnackbarItem({ message: 'Copied to clipboard' });
}).catch((e) => { console.log(e); });
```

Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: Sounds good :D I did have to import `addSnackbarItem` though. :+1: But this should be resolved :slightly_smiling_face:

**Comment by nizarmah** [comment]
> However, while testing this I found out that my PR introduced a bug in the comment permalink page. Would you mind including this change in your PR?

@fdoxyz ofcourse I don't mind :slightly_smiling_face: Let me know if there's anything else you'd want me to change :smile:

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (fdoxyz) [inline app/assets/javascripts/initializers/initializeCommentDropdown.js:2]: No need to import here. This is an initializer that runs in a global scope

**Review by fdoxyz — CHANGES_REQUESTED** [review]
Just one line to remove

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: Oh! Sorry about that! eslint was showing an error so I thought of importing it. I wasn't aware that this runs in a global scope (I did see the comment up top, but I wasn't certain if I had to suppress the eslint error)

Anyway, I removed it and suppressed the eslint error :+1: Sorry again!

Comment (fdoxyz) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: Oh sorry, I forgot that you were going to run into a eslint check because `addSnackbarItem` wasn't declared in this file. The way to properly fix the eslint is to add it to that annotation at the very top so it would look like this:

```js
/* global Runtime, addSnackbarItem */
```

**Review by fdoxyz — APPROVED** [review]
It's working great @nizarmah. There may be a small delay in getting another approval as the team will be a bit [slower to respond this coming week](https://dev.to/devteam/the-dev-forem-team-is-taking-a-mental-health-holiday-1ldl), but this LGTM now.

**Comment by nizarmah** [comment]
> There may be a small delay in getting another approval as the team will be a bit slower to respond this coming week

No worries, I totally understand. This year has been tough on everyone :slightly_frowning_face: I hope you guys enjoy that time off and find a way to take a break from the crazy events of this year :hugs:

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: Would you like me to change the `// eslint-disable-next-line`? And replace it with the fix you suggested above?

**Comment by nizarmah** [comment]
Hope you guys are doing good :slightly_smiling_face: Any updates regarding this PR @fdoxyz ?

**Comment by nizarmah** [comment]
@fdoxyz unfortunately, I'm not much of a React person yet :( But I can definitely learn if there's a need for it. Thanks for the update though! And I'll be waiting on the review, to know when exactly the changes need to be changed to the form of a React component. :+1:

**Comment by mstruve** [comment]
@nickytonline any thoughts on this?

**Comment by nickytonline** [comment]
This one flew under the radar for me. Thanks for the bump @mstruve. I will take a peek at this this week.

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeCommentDropdown.js:41]: Refactor (blocking): This catch is not necessary because if this fails, it's an unexpected exception and Honeybadger (our reporting tool), will not be able to log it as the `console.log` will swallow the error.

By default, Honeybadger handles unhandledrejections. See https://github.com/honeybadger-io/honeybadger-js/blob/bc074bc335164d40655489d017bfcd9f0f64a36d/src/browser/integrations/onunhandledrejection.ts#L9

**Review by nickytonline — CHANGES_REQUESTED** [review]
This looks great. Thanks so much for being patient on this one. One small change request and then this is good to merge.

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: Alright, thanks for that note, I've removed the catch statement. :+1: 
I'll keep this comment unresolved so that it'll be easier to notice when checking the changes.
Please let me know if there's anything else :smile:

**Comment by nizarmah** [comment]
@fdoxyz would you be able to help me please re-run the Travis CI build, the option isn't showing up for me.
Also, if you have any idea about any comment message I can send to trigger it, please let me know :+1:

**Comment by fdoxyz** [comment]
A restart worked. Tests are green again 👍

**Comment by nizarmah** [comment]
Thanks a lot @fdoxyz :partying_face: 

@nickytonline this is ready for your review again :smile:

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeCommentDropdown.js:141]: Refactor (non-blocking): adding `ButtonListener` is redundant

**Review by nickytonline — APPROVED** [review]
Thanks so much for the PR! 😎

File: app/assets/javascripts/initializers/initializeCommentDropdown.js
Comment (nizarmah) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js:141]: @nickytonline would you prefer I changed it to `copyPermalinkListener` only?

```suggestion
  function copyPermalinkListener(copyPermalinkButton) {
```

Comment (nickytonline) [reply app/assets/javascripts/initializers/initializeCommentDropdown.js]: No that's fun and thanks again!

**Comment by nizarmah** [comment]
Thanks a lot @fdoxyz and @nickytonline for your review! It's awesome contributing to such an awesome community and platform. :heart: Your team is amazing, have a wonderful week!

**Pull request merged**
